nilearn/regions/region_extractor.py

```diff
@@ -0,0 +1,313 @@
+
+    Returns
+    -------
+    threshold_maps_img: Nifti1Image object
+        gives us thresholded image.
+    """
+    maps = check_niimg(maps_img)
+    n_maps = maps.shape[-1]
+    if not isinstance(threshold, float) or threshold <= 0 or threshold > maps.shape[-1]:
+        raise ValueError("threshold given as ratio to the number of voxels must "
+                         "be float value and should be positive and between 0 and "
+                         "total number of maps i.e. n_maps={0}. "
+                         "You provided {1}".format(maps.shape[-1], threshold))
+    else:
+        ratio = threshold
+
+    maps_data = maps.get_data()
+    abs_maps = np.abs(maps_data)
+    # thresholding
+    cutoff_threshold = scoreatpercentile(
+        abs_maps, 100. - (100. / n_maps) * ratio)
+    maps_data[abs_maps < cutoff_threshold] = 0.
+
+    threshold_maps_img = new_img_like(maps, maps_data)
+
+    return threshold_maps_img
+
+
+def connected_regions(maps_img, min_region_size=50, extract_type='local_regions',
+                      smoothing_fwhm=6, mask_img=None):
+    """ Extraction of brain connected regions into separate regions.
+
+    Parameters
+    ----------
+    maps_img: Niimg-like object
+        an image of brain activation or atlas maps to be extracted into set of
+        separate brain regions.
+
+    min_region_size: int, default 50, optional
+        Minimum number of voxels for a region to be kept. Useful to suppress
+        small spurious regions.
+
+    extract_type: str {'connected_components', 'local_regions'} \
+        default local_regions, optional
+        If 'connected_components', each component/region in the image is extracted
+        automatically by labelling each region based upon the presence of unique
+        features in their respective regions.
+        If 'local_regions', each component/region is extracted based on their
+        maximum peak value to define a seed marker and then using random walker
+        segementation algorithm on these markers for region separation.
+
+    smoothing_fwhm: scalar, default 6mm, optional
+        To smooth an image to extract most sparser regions. This parameter
+        is passed `_smooth_array` and exists only for extract_type 'local_regions'.
+
+    mask_img: Niimg-like object, default None
+        If given, mask image is applied to input data.
+        If None, no masking is applied.
+
+    Returns
+    -------
+    regions_extracted_img: Nifti1Image object
+        gives the image in 4D of extracted brain regions. Each 3D image consists
+        of only one separated region.
+
+    index_of_each_map: numpy array
+        an array of list of indices where each index denotes the identity
+        of each extracted region to their family of brain maps.
+    """
+    all_regions_imgs = []
+    index_of_each_map = []
+    maps_img = check_niimg(maps_img, atleast_4d=True)
+    maps = maps_img.get_data()
+    affine = maps_img.get_affine()
```

Comment:
but... (affine[0]* affine[1] * affine[2]) is a vector... (=0 most of the time...)